src/rapid_pro_to_engagement_db/rapid_pro_to_engagement_db.py

```diff
@@ -0,0 +1,155 @@
+    filter = lambda q: q \
+        .where("text", "==", message.text) \
+        .where("timestamp", "==", message.timestamp) \
+        .where("participant_uuid", "==", message.participant_uuid)
+
+    matching_messages = engagement_db.get_messages(filter=filter)
+    assert len(matching_messages) < 2
+
+    return len(matching_messages) > 0
+
+
+def _ensure_engagement_db_has_message(engagement_db, message, message_origin_details):
+    """
+    Ensures that the given message exists in an engagement database.
+
+    This function will only write to the database if a message with the same text, timestamp, and participant_uuid
+    doesn't already exist in the database.
+
+    :param engagement_db: Engagement database to use.
+    :type engagement_db: engagement_database.EngagementDatabase
+    :param message: Message to make sure exists in the engagement database.
+    :type message: engagement_database.data_models.Message
+    :param message_origin_details: Message origin details, to be logged in the HistoryEntryOrigin.details.
+    :type message_origin_details: dict
+    """
+    if _engagement_db_has_message(engagement_db, message):
+        log.debug(f"Message already in engagement database")
+        return
+
+    log.debug(f"Adding message to engagement database")
+    engagement_db.set_message(
+        message,
+        HistoryEntryOrigin(origin_name="Rapid Pro -> Database Sync", details=message_origin_details)
+    )
+
+
+def sync_rapid_pro_to_engagement_db(google_cloud_credentials_file_path, rapid_pro_to_engagement_db_configuration,
+                                    engagement_db, uuid_table):
+    """
+    Synchronises runs from a Rapid Pro workspace to an engagement database.
+
+    :param google_cloud_credentials_file_path: Path to a Google Cloud service account credentials file to use to access
+                                               the credentials bucket.
+    :type google_cloud_credentials_file_path: str
+    :param rapid_pro_to_engagement_db_configuration: Configuration for the sync operation.
+    :type rapid_pro_to_engagement_db_configuration: src.rapid_pro_to_engagement_db.configuration.RapidProToEngagementDBConfiguration
+    :param engagement_db: Engagement database to sync to.
+    :type engagement_db: engagement_database.EngagementDatabase
+    :param uuid_table: UUID table to use to de-identify contact urns.
+    :type uuid_table: id_infrastructure.firestore_uuid_table.FirestoreUuidTable
+    """
+    # This implementation is WIP. It shows how we can non-incrementally synchronise a workspace to the database.
+    # To enter production, we still need the following:
+    # TODO: Support incremental update of runs and contacts.
+    # TODO: Handle deleted contacts.
+    # TODO: Handle contacts that have runs but haven't been fetched locally yet.
+    # TODO: Optimise fetching fields from the same flows, so we don't have to download the same runs multiple times.
+
```

Comment:
TODO: Auto code before adding to database

Reply:
I think this is best done when syncing to Coda, as per the design doc. Then all the logic and configuration related to labelling is all done in one place, the coda sync.